### ISSUE TITLE
Migrations on overlapping plans can run concurrently.

### ISSUE THREAD

**Issue by jortel** [fusor/mig-controller#172]
Overview
---

Foreach migration plan, the controller runs migrations serially.  The goal is to prevent migrations from doing conflicting things on the clusters involved or otherwise allowing the migrations to interfere with each other.  However, I don't think this approach alone is sufficient.  Currently, multiple plans can be created for the same source and destination clusters, and, with overlapping namespaces.  This defeats the migration (run) orchestration.  Running 2 migrations concurrently on different plans with the same source and destination clusters is effectively the same as running 2 migrations concurrently on the same plan.

It's my understanding that the use case for multiple plans with the same source & destination cluster is to support migrating workloads by groups of namespaces in sequential passes/efforts.

The conflict
---

Let's start with a complete understanding of the _conflict_. 
- The source cluster overlap will result in a Restic pod restart to be triggered by one migration in the middle of another migration's backup.  With overlapping namespaces: the pod quiesce will conflict; duplicate stage pods will be created/deleted; The PV copy/move will likely conflict.
- The destination cluster overlap with overlapping namespaces: the stage pods will conflict; The PV copy/move will conflict (or just overlap?).

Proposed Solution
---

There are 2 options for preventing this (that I can think of).

`Option 1` seems like the best option assuming it does not block any known use cases.

**Option 1:**  Add a validation that will block running migrations on (open) plans that have the same source or destination cluster.  The validation would set a _Conflict_ condition that will prevent the _Ready_ condition which will block migrations on overlapping plans.  Would this block any known user cases?

Pros:
- Simple.
- Alerts the user to the overlap.
- Seems more intuitive for the user.
- Keeps complexity low in the migration (run) orchestration algorithm.
- Promotes work flow of closing completed plans.

Cons:
- _May_ give the user impression that overlapping plans are unhealthy when they're not.
- _May_ block a use case I'm not aware of.

**Option 2:**  Enhance the migration (run) orchestration to run migrations serially across plans that have the same source or destination cluster.

Pros:
- Migrations will seem to _sort them selves out_ and just eventually run.
- Overlapping plans will remain Ready which give the user the feeling that things are healthy.

Cons:
- Migration (run) orchestration across plans adds a good bit of complexity.
- User may get confused about why the migrations on a plan are being postponed prompting question s like _"When will my migration run?"_
- The changes (edits) on one plan, will block/interrupt migrations running on another plan.
- User not alerted to the overlap.
- Does not promote plan closing work flow.
- Requires the (run) ordering of overlapping migration plans.  The order created?   This fixed ordering could conflict with user intentions.  Then what?
 
Related to #156

**Comment by jwmatthews** [comment]
I'm all for an approach of being safe in first release, restrict what we need to minimize problems.  

Need to learn a little more on what the experience of Option 1 would be. 

In subsequent releases would like to see if we could allow operations in parallel that didn't involve same namespaces, i.e. would be nice if the granularity was lower than just same source_cluster/destination_cluster, but if it went a level deeper and looked at the specific namespaces in each to determine if overlap will occur.

Assume we will need to account for case of multiple migration controllers on other systems using same source/destination, so need some kind of mechanism of maybe 'locking' a namespace we are operating on so another instance of the migration controller is aware an operation is underway.

In terms of 4.2 release, I think it's best to limit complexity and nail down a safe route, then come back to allowing parallel runs and better logic in next release.

As to restarting Restic, I was under impression that was only for OCP 3.7 and 3.9 which lacked mount propagation, would assume we wouldn't want to restart Restic on 3.10+ clusters.

**Comment by jortel** [comment]
> I'm all for an approach of being safe in first release, restrict what we need to minimize problems.
> 
> Need to learn a little more on what the experience of Option 1 would be.
> 
> In subsequent releases would like to see if we could allow operations in parallel that didn't involve same namespaces, i.e. would be nice if the granularity was lower than just same source_cluster/destination_cluster, but if it went a level deeper and looked at the specific namespaces in each to determine if overlap will occur.

I think the condition could factor in overlapping namespaces (in addition to source & destination cluster) and give you the granularity you want provided the Restic Pod restart in the middle of a Backup would not be harmful.

> 
> Assume we will need to account for case of multiple migration controllers on other systems using same source/destination, so need some kind of mechanism of maybe 'locking' a namespace we are operating on so another instance of the migration controller is aware an operation is underway.

Multiple controllers working on the same source and destination clusters introduces race conditions into the proposed _Conflict_ validation AND the migration serialization.  If this is a requirement, we have a difficult problem to solve.

> 
> In terms of 4.2 release, I think it's best to limit complexity and nail down a safe route, then come back to allowing parallel runs and better logic in next release.
> 
> As to restarting Restic, I was under impression that was only for OCP 3.7 and 3.9 which lacked mount propagation, would assume we wouldn't want to restart Restic on 3.10+ clusters.

Agreed.  Although, the controller currently has a TODO to make that decision. The validation logic could account for the OCP version and be more liberal when possible.

**Comment by jwmatthews** [comment]
> > I'm all for an approach of being safe in first release, restrict what we need to minimize problems.
> > Need to learn a little more on what the experience of Option 1 would be.
> > In subsequent releases would like to see if we could allow operations in parallel that didn't involve same namespaces, i.e. would be nice if the granularity was lower than just same source_cluster/destination_cluster, but if it went a level deeper and looked at the specific namespaces in each to determine if overlap will occur.
> 
> I think the condition could factor in overlapping namespaces (in addition to source & destination cluster) and give you the granularity you want provided the Restic Pod restart in the middle of a Backup would not be harmful.
> 
> > Assume we will need to account for case of multiple migration controllers on other systems using same source/destination, so need some kind of mechanism of maybe 'locking' a namespace we are operating on so another instance of the migration controller is aware an operation is underway.
> 
> Multiple controllers working on the same source and destination clusters introduces race conditions into the proposed _Conflict_ validation AND the migration serialization. If this is a requirement, we have a difficult problem to solve.
> 

We don't have an explicit requirement for supporting multiple controllers.  Just some thoughts below of why I thought we might need to consider a solution that was beyond any view a single instance of our controller might have.

The thought was, if we allowed a consultant to install the migration controllers/api/webui on their laptop and do a migration between an external source and destination cluster....I could see this being extended to multiple consultants installing the same on their own laptops, then multiple controllers migrating workloads, hence a need to do something like 'lock' a namespace, place a ConfigMap in namespace with a known name to share state or something.

There is also the situation that it's possible a destination cluster may have several source clusters fanning into it.....I don't think there will need to be a 1-to-1 mapping of source to destination clusters.

Could see various patterns of a source cluster being used to copy data from it to multiple destinations and all the permutations that go with this, multiple sources to one destination.

I also see that some situations we'd want the migration controller to be on destination and in others we might want it to be on source, depending on if we are migrating from on-prem to cloud or similar environments where directionality of requests and network traffic matter.

Last thought is in next release we would like to open up access to beyond cluster-admin, so this would be a tool an end user could run, not just a cluster-admin...when that happens I think we need to be careful of cluster wide things....and try to constrain needs down to the namespace level.

Requirements are not fleshed out, just trying to think through some things that might be needed in near future.



> > In terms of 4.2 release, I think it's best to limit complexity and nail down a safe route, then come back to allowing parallel runs and better logic in next release.
> > As to restarting Restic, I was under impression that was only for OCP 3.7 and 3.9 which lacked mount propagation, would assume we wouldn't want to restart Restic on 3.10+ clusters.
> 
> Agreed. Although, the controller currently has a TODO to make that decision. The validation logic could account for the OCP version and be more liberal when possible.

**Comment by jortel** [comment]
> The thought was, if we allowed a consultant to install the migration controllers/api/webui on their laptop and do a migration between an external source and destination cluster....I could see this being extended to multiple consultants installing the same on their own laptops, then multiple controllers migrating workloads, hence a need to do something like 'lock' a namespace, place a ConfigMap in namespace with a known name to share state or something.

What is the purpose of having each consultant run their own controller on their laptop?  To avoid installing the controller/CRDs on the source or destination cluster?  Would each laptop be running the controller in it's own cluster?

> There is also the situation that it's possible a destination cluster may have several source clusters fanning into it.....I don't think there will need to be a 1-to-1 mapping of source to destination clusters.

Yes, I anticipated this.
 
> Could see various patterns of a source cluster being used to copy data from it to multiple destinations and all the permutations that go with this, multiple sources to one destination.

Yes, I anticipated this.

> 
> Last thought is in next release we would like to open up access to beyond cluster-admin, so this would be a tool an end user could run, not just a cluster-admin...when that happens I think we need to be careful of cluster wide things....and try to constrain needs down to the namespace level.

A set of namespace seems to be the natural boundary.

**Comment by jwmatthews** [comment]
> > The thought was, if we allowed a consultant to install the migration controllers/api/webui on their laptop and do a migration between an external source and destination cluster....I could see this being extended to multiple consultants installing the same on their own laptops, then multiple controllers migrating workloads, hence a need to do something like 'lock' a namespace, place a ConfigMap in namespace with a known name to share state or something.
> 
> What is the purpose of having each consultant run their own controller on their laptop? To avoid installing the controller/CRDs on the source or destination cluster? Would each laptop be running the controller in it's own cluster?
> 

I don't know of a concrete use case where this is required.  The purpose came about more to help us think of a solution that was flexible when we first thought of modeling what is needed for the WebUI in terms of how to talk to multiple clusters.  Thought was if we could support being installed on a third party laptop and have it be the executor, then we have flexibility for unknowns in future.


> > There is also the situation that it's possible a destination cluster may have several source clusters fanning into it.....I don't think there will need to be a 1-to-1 mapping of source to destination clusters.
> 
> Yes, I anticipated this.

This looks related to Issue #47 

> 
> > Could see various patterns of a source cluster being used to copy data from it to multiple destinations and all the permutations that go with this, multiple sources to one destination.
> 
> Yes, I anticipated this.
> 
> > Last thought is in next release we would like to open up access to beyond cluster-admin, so this would be a tool an end user could run, not just a cluster-admin...when that happens I think we need to be careful of cluster wide things....and try to constrain needs down to the namespace level.
> 
> A set of namespace seems to be the natural boundary.

**Comment by jwmatthews** [comment]
I saw UI issue that may be related to this: https://github.com/fusor/mig-ui/issues/328

I created multiple overlapping MigPlans, then I saw how running a Stage or Migration on one would update the status of others.  What I saw I think was an issue on ui side and not related to controller, but adding a comment here for help tracking

@djwhatle will be investigating this on UI side in coming week.
